src/Microsoft.AspNet.Server.Kestrel.Https/HttpsConnectionFilter.cs

```diff
@@ -35,8 +42,72 @@ public HttpsConnectionFilter(X509Certificate2 cert, IConnectionFilter previous)
+                SslStream sslStream;
+                if (_options.ClientCertificateMode == ClientCertificateMode.NoCertificate)
+                {
+                    sslStream = new SslStream(context.Connection);
+                    await sslStream.AuthenticateAsServerAsync(_options.ServerCertificate, clientCertificateRequired: false,
+                        enabledSslProtocols: _options.SslProtocols, checkCertificateRevocation: _options.CheckCertificateRevocation);
+                }
+                else
+                {
+                    sslStream = new SslStream(context.Connection, leaveInnerStreamOpen: false,
+                        userCertificateValidationCallback: (sender, certificate, chain, sslPolicyErrors) =>
+                        {
+                            if (certificate == null)
+                            {
+                                return _options.ClientCertificateMode != ClientCertificateMode.RequireCertificate;
+                            }
+
+                            if (_options.ClientCertificateValidation == null)
+                            {
+                                if (sslPolicyErrors != SslPolicyErrors.None)
+                                {
+                                    return false;
+                                }
+                            }
+
+                            X509Certificate2 certificate2 = certificate as X509Certificate2;
+                            if (certificate2 == null)
+                            {
+#if DOTNET5_4
+                                // conversion X509Certificate to X509Certificate2 not supported
+                                // https://github.com/dotnet/corefx/issues/4510
+                                return false;
+#else
+                                certificate2 = new X509Certificate2(certificate);
+#endif
+                            }
+
+                            if (_options.ClientCertificateValidation != null)
+                            {
+                                if (!_options.ClientCertificateValidation(certificate2, chain, sslPolicyErrors))
+                                {
+                                    return false;
+                                }
+                            }
+
+                            clientCertificate = certificate2;
+                            return true;
+                        });
+                    await sslStream.AuthenticateAsServerAsync(_options.ServerCertificate, clientCertificateRequired: true,
+                        enabledSslProtocols: _options.SslProtocols, checkCertificateRevocation: _options.CheckCertificateRevocation);
+                }
+
+                var previousPrepareRequest = context.PrepareRequest;
+                context.PrepareRequest = features =>
+                    {
+                        previousPrepareRequest?.Invoke(features);
+
+                        if (clientCertificate != null)
```

Comment:
It would be nicer if the code could use SslStream.RemoteCertificate instead of this variable. But RemoteCertificate isn't the same instance as the one passed to the validation callback. The validation callback gets an X509Certificate2 while RemoteCertificate is an X509Certificate.